ADR/ADR00TBD-use-rust-for-api.md

```diff
@@ -0,0 +1,52 @@
+
+Alongside this, we also have to consider the supportability and maintainability of this API with respect to the tech team who will need to develop and support this API. This must consider the cost of hiring or training developers in Rust, as well as what support is available more widely should it be necessary in the future. As Rust is still a relatively new language (and not one officially adopted within the GDS) these aspects to our decision making are similarly important to the technological choice.
+
+### Rust Advantages
+Rust bills itself as _"A language empowering everyone to build reliable and efficient software"_. It was first released in 2010, 8 months after Go. In terms of adoption there are [about the same number](https://redmonk.com/sogrady/files/2021/08/lang.rank_.0621.png) of github projects written in Rust as there are in Go.
+
+Rust is designed to quickly build provably correct software, with an emphasis on zero runtime failures, and high-quality developer ergonomics.
+
+In alpha testing, we wrote a prototype API, powering the X-Gov form builder UI in three languages, the source code for these three prototypes are available in branches of [alphagov/forms-api-prototype](https://github.com/alphagov/forms-api-prototype):
+ 1. [Ruby](https://github.com/alphagov/forms-api-prototype/pull/1)
+ 2. [Rust](https://github.com/alphagov/forms-api-prototype/pull/2)
+ 3. [Python](https://github.com/alphagov/forms-api-prototype/pull/3)
+
+In our prototyping, we saw that Rust is a very good match for the project we planned on building, despite it being a new language to the GDS.
+
+Notable benefits are:
+- An [76x decrease](https://github.com/drujensen/fib#results) in execution times (compared to Python, 32x decrease compared to Ruby), and a consummate reduction in power requirement (and therefore carbon cost)
+	- Request round-trip times are also faster: The slowest Rust web frameworks [are faster](https://web-frameworks-benchmark.netlify.app/result?asc=0&l=rust,ruby,python]) than the fastest Python and Ruby frameworks. Ruby is 1.45x slower, Python, 1.85x. (only the [sifrr](https://sifrr.github.io/sifrr/#/./packages/server/sifrr-server/) framework is faster in JS land)
+- A up to an order of magnitude reduction in memory overhead compared to python and ruby (depending on application) due to lack of runtime overhead.
+- Faster development due to instant compiler feedback (instead of waiting for logs)
+- Faster development due to less unit testing burden: Many errors that would require unit testing are instead compile-time errors. This is not just simple type errors, but whole categories of invalid states are unrepresentable due to careful language and library design.
+- 100% compiled coverage by definition (all compiled code is valid). Allowing the team great confidence that no API responses are invalid, no inputs cause errors, etc.
```

Comment:
There's something that feels off about these three, they feel a little subjective I think. Whilst it may be quicker for some to face compile errors, it may also be less quick for others depending on their particular way of approaching problems.

I'd rather condense these down to a short summary highlighting something more around the what the final point touches on - with the compiler giving a lot more confidence about written code given its feature set over discussion potential development speed benefits.

Similarly, though it's touched on in the "training" section of disadvantages, it may also be worth calling out there the counterpart to the compiler being so complex, which is that it creates an incredibly large initial learning curve where you're likely to need to understand a lot about the compiler to be able to get going with Rust - something that could run the risk of being very challenging to someone more junior as there's a lot to understand before your code even starts to run.

Reply:
Agreed, rewritten